Created make_level script to automatically generate new level files

parkmichael777 opened the pull request:
fixes #102

parkmichael777 commented:
heyo, I pushed the finished version. It's a little jank so let me know if I should change something.

parkmichael777 commented:
@bthayer2365 just to make sure you saw this

bthayer2365 commented:
Thanks for the reminder! Just put my head down for the last few hours trying to get the test suites up and running. It's done now, pull master and check it out!

I'll review this now

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,104 @@
+join = os.path.join
+cwd = os.getcwd()
+
+def writeInit(skill_name, skill_path, level_name):
```
call this `write_init`

parkmichael777 replied:
Okay, this change will be reflected when I push

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,104 @@
+    # If skills/skill/__init__.py doesn't exist, create a basic version (no levels)
+    if not path.exists(join(skill_path,"__init__.py")):
+        with open(join(skill_path,"__init__.py"), 'w+') as fp:
+            level_setup = "import pkg_resources\nfrom gitgud.skills.util import BasicLevel\nfrom gitgud.skills.util import Skill\n\nskill = Skill(\n\t'{}',\n\t[\n\t]\n)".format(skill_name)
```
Separate this line out into multiple lines

```
level_setup = '\n'.join([
    "stuff",
    "morestuff",
    "", # blank line
])
```

bthayer2365 replied:
It might also be worth it to just read in a template file that is already all structured out, and then you just fill in the blanks with the brackets after reading it into a variable

bthayer2365 replied:
You can do `string = string.replace('\t', ' ' *4)`

parkmichael777 replied:
I replaced tabs with ' ' * 4 using the method described above, split up some of the string concatenations, but left some (the ones that only write one line to a file) let me know if that needs to be changed further.

bthayer2365 commented:
I was reviewing that whole time but it was worth it. This is going to be very useful... 

After those changes, I'll look deeper into the code and let you know about anything else

bthayer2365 commented on gitgud/make_level.py:
```diff
@@ -0,0 +1,114 @@
+    level_name = sys.argv[2]
+    
+    # Check if cwd is git-gud folder
+    if cwd[-7:] == "git-gud":
```
I see your new change. Let me clarify, I actually think it's good for the file to be in the top level directory. The change I was referring to was on this line. Just check if there is `os.path.isdir(os.path.join(cwd, 'gitgud'))`

bthayer2365 replied:
Can you move it back and change this line?

parkmichael777 replied:
Changed

parkmichael777 replied:
some bugs hold off on review

parkmichael777 commented:
Regarding the changes requested above:
1. Which __init__.py? And do you mean at the very end of the file?
2. Added newline after pkg_resources
3. Tabs replaced with four spaces. Originally used .replace("\t", " " * 4) but this caused problems so just ctrF and replaced the \t's with "    ". fight me lol
4. I am not sure what the test cases are. Let me know what they do/where they are and how they interact with this stuff.
5. Is this for creating the empty text files? Could you explain this syntax a little more?
6. Contingent on 5.
7. Re-added os.path and os.path.join
8. Changed prompt
9. Done-zo
10. That wasn't actually supposed to be there lol, I was just using that to check that the file pointers closed. I'm not sure if python magic takes care of that for you and I didn't want a memory leak lmao.
11. By logs, do you mean display on the terminal whether files were successfully created?
12. Fixed!

bthayer2365 commented:
Regarding the changes requested above:
1. Which __init__.py? And do you mean at the very end of the file?
 - I mean `gitgud/skills/blah/__init__.py` at the end
2. Added newline after pkg_resources
 - great
3. Tabs replaced with four spaces. Originally used .replace("\t", " " * 4) but this caused problems so just ctrF and replaced the \t's with "    ". fight me lol
 - o.0
4. I am not sure what the test cases are. Let me know what they do/where they are and how they interact with this stuff.
 - I think it's self-explanatory, but let me know. They're in each skill folder, so there's one in `gitgud/skills/basics/test_levels.py`
5. Is this for creating the empty text files? Could you explain this syntax a little more?
 - It creates a file for writing. 'w+' means write ('w') and create if it doesn't exist ('+')
 - What you had before, 'a+' means append and create if it doesn't exist
 - The with syntax uses a closable type, it's the equivalent of 
```
file = open(file_name, 'w+')
pass
file.close()
```
6. Contingent on 5.
7. Re-added os.path and os.path.join
8. Changed prompt
9. Done-zo
 - 💯 
10. That wasn't actually supposed to be there lol, I was just using that to check that the file pointers closed. I'm not sure if python magic takes care of that for you and I didn't want a memory leak lmao.
11. By logs, do you mean display on the terminal whether files were successfully created?
 - Basically. Just let me know every time something I care about happens. "File X created" "Level added in Y" "Test created" etc
12. Fixed!
 - 👍

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,119 @@
+
+cwd = os.getcwd()
+
+def write_Init(skill_name, skill_path, level_name):
```
nO CaPs!!

parkmichael777 replied:
haha changed

bthayer2365 commented on make_level.py:
```diff
@@ -31,33 +30,29 @@ def write_Init(skill_name, skill_path, level_name):
 
-        replace = ",\n\t\tBasicLevel('{0}', pkg_resources.resource_filename(__name__, '_{0}/'))\n\t]".format(level_name)
-        filedata = filedata.replace("\n\t]", replace)
+        replace = ",\n        BasicLevel('{0}', pkg_resources.resource_filename(__name__, '_{0}/'))\n    ]".format(level_name)
```
replace = ",\n        BasicLevel('{level_name}', pkg_resources.resource_filename(__name__, '_{level_name}/'))\n    ]".format(level_name=level_name)

bthayer2365 commented:
I've got something else fun for you to do. 
```
pip install flake8
flake8 make_level.py
```

bthayer2365 commented:
Google pep8. It's a standard that tries to make code more readable and consistent. It has all kinds of things from spacing standards to what to name your variables and such. Flake8 tells you where you violate pep8.

bthayer2365 commented on make_level.py:
```diff
@@ -55,6 +55,11 @@ def write_Init(skill_name, skill_path, level_name):
+def create_file(level_path,filename):
+    with open(os.path.join(level_path,filename),'a+'):
+        pass
+    return
```
No return statement neccessary

parkmichael777 replied:
changed

bthayer2365 commented:
Let me know when you're ready for a review

parkmichael777 commented:
Everything should be done except the test cases. I was confused because I forgot to pull. I'll figure out what they do and look up pep8 my next worktime. Today, I'm probably gonna stop here.

bthayer2365 commented:
Sounds good. I'll get the review going and then you can continue whenever. If you don't understand the test cases, just ask

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,117 @@
+                os.mkdir(level_path)
+                print("Created: {}".format(level_name))
+            
+            print("\nModifying Init Files:")
```
Let's say "Registering level" then log separately file creation and modification.
Registering level "foo bar"
Registered skill "foo" in "gitgud\skills\__init__.py"
Creating file "gitgud\skills\foo\__init__.py"
Registering level "bar" in "gitgud\skills\foo\__init__.py"

parkmichael777 replied:
Changed

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,117 @@
+            skill_path = os.path.join("gitgud","skills","{}".format(skill_name))
+            if not os.path.exists(skill_path):
+                os.mkdir(skill_path)
+                print("Created: {}".format(skill_name))
```
Let's output the path, so 
Created: "gitgud/skills/foo/"
Created: "gitgud/skills/foo/_bar/"

parkmichael777 replied:
Reflected in next push

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,117 @@
+            print("Modified: Added level to {}".format(filepath))
+    return
+
+def create_file(level_path,filename):
```
Can we call this `create_level_file` since it's not just creating any file, it's the file for the level

parkmichael777 replied:
Done

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,117 @@
+    
+        
+        # Read file
+        with open(os.path.join(os.path.join("gitgud","skills"),"__init__.py"), 'r') as fp:
```
I think you can do 
```
os.path.join("gitgud", "skills", "__init__.py")
```

parkmichael777 replied:
Lol yeah I used this format in other parts of the code haha idk why I started nesting them along the way. Fixed

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,117 @@
+                "    [",
+                "    ]",
+                ")\n",
+                ])
```
For all of these, instead of newlines, can you do a new entry with a blank line, so
```
                ")\n",
```
Turns into
```
                ")",
                "",
```
I think it makes it clear, and that way we can see the visual separation here too

parkmichael777 replied:
Do you want this for the one line statements too?

bthayer2365 replied:
Yeah, we should be able to see what the files look like instead of just having the most compressed form

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,117 @@
+def main():
+    # Obtain input arguments
+    skill_name = sys.argv[1]
+    level_name = sys.argv[2]
```
Argument handling. try/except with error message telling correct usage

parkmichael777 replied:
I think I did it? New to python so I've never used this. Let me know.

parkmichael777 commented:
Addressed all separate comments above and the newline issue you mentioned in the change request.

Still need to do test case generation, and look at flake8/template files. That's it for now

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,128 @@
+            if choice == 'n':
+                return
+            
+            print("\nCreating Folders:")
```
For here and down, you can change the newlines to prints, just so it's a little more visually appealing. Also, like to keep the print statements after each line of text, so here:
```
print()
print("Creating folders")
print()
```

parkmichael777 replied:
done

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,128 @@
+                os.mkdir(level_path)
+                print("Created: {}".format(level_path))
+            
+            print("\nRegistering Level: {} {}".format(skill_name, level_name))
```
Here, it would be
```
print("Registering level...")
print()
```

parkmichael777 replied:
done

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,128 @@
+        skill_name = sys.argv[1]
+        level_name = sys.argv[2]
+    except:
+        print("Usage: \"python make_level.py <skill_name> <level_name>\"")
```
Add a return statement here

parkmichael777 replied:
done

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,128 @@
+        print("Usage: \"python make_level.py <skill_name> <level_name>\"")
+        
+    # Check if current dir is ../gitgud directory. (i.e. dir of setup files)
+    if os.path.isdir(os.path.join(cwd, 'gitgud')):
```
Instead of doing an if/else, I think it makes sense to do an if/return and keep the rest of the code unindented, so 
```
if not os.path.isdir(...):
    print(stuff)
    return
# Rest of the code below will go here and will have one less indent
```

parkmichael777 replied:
done

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,128 @@
+            create_level_file(level_path, "test.txt")
+    else:
+        print("Error: Execute this script in the git-gud directory.")
+
```
Once the function is over, can we add a `print("Done.")`

parkmichael777 replied:
done

bthayer2365 commented on make_level.py:
```diff
@@ -0,0 +1,128 @@
+    filepath = os.path.join(level_path,filename)
+    with open(filepath,'a+'):
+        pass
+    print("Created: {}".format(filepath))
```
Let's make one file each for each and use them as templates for each of the files. The templates should explain what's going on and still work, so the .spec files should be valid (use tag and branch names to imply what each thing does), goal.txt should be short, and instructions.txt should use >>> and explain itself on multiple lines.

Once you do that, have this function copy those files over to the appropriate directory. It might make sense to look for a copy function in the os module.

parkmichael777 replied:
Done

bthayer2365 commented:
Hey, we're also going to need to add an entry to setup.py to register the package

parkmichael777 commented:
For the setup thing do i just add it to packages?

benthayer commented:
No, more like
```
packages=[
    'gitgud',
    'gitgud.skill.newskill'
],
```

benthayer commented:
make_level.py is just a utility for us as developers, not something we need to deploy to everyone

parkmichael777 commented:
Ok, can you explain what that does?

This code you wrote:
packages=[
    'gitgud',
    'gitgud.skill.newskill'
],

benthayer commented:
Yeah, this is not something that you will notice, only something that matters when I'm deploying the code and someone installs it. When you're running locally, python just looks through the filesystem and sees that the packages are there, but for example, there may be tests we don't want to include in the package, so we don't specify it as a module. When we specify it, python's setuptools knows which packages to include

benthayer commented:
If you pull the most recent version of the code, you'll notice that there's also 
```
    package_data={
        'gitgud.skills.basics': ['_*/*'],
        'gitgud.skills.extras': ['_*/*'],
        'gitgud.skills.rampup': ['_*/*'],
    },
```
You'll need to add this too

benthayer commented on level_file_temp/setup.spec:
```diff
@@ -0,0 +1,9 @@
+4 : 1 (Example_child_of_1)
+5 (tag : Example_child_of_4)
+6
+7 : 3 6 (Example_merge_node_of_3_&_4)
```
If we use &, then we would have to always put the tag name in quotes because & is treated as a special operator, so it might be better to use `merge-3-and-4` or something else

parkmichael777 replied:
done

parkmichael777 commented on level_file_temp/goal.txt:
```diff
@@ -0,0 +1 @@
+In this file, explain to the user/player what the goal of the level is (e.g. "Merge commits 1 and 2 to complete the level!").
```
done

parkmichael777 commented on level_file_temp/instructions.txt:
```diff
@@ -0,0 +1,3 @@
+Use this file to explain to the users how to use the git feature that this level focuses on.
+>>>
+Separate blocks using ">>>", and be sure to include tips, and examples as explanation aids.
```
done

parkmichael777 commented on level_file_temp/setup.spec:
```diff
@@ -0,0 +1,9 @@
+1 (tag : <tag_name>)
```
done

parkmichael777 commented on level_file_temp/setup.spec:
```diff
@@ -0,0 +1,9 @@
+1 (tag : <tag_name>)
```
done

parkmichael777 commented on level_file_temp/setup.spec:
```diff
@@ -0,0 +1,9 @@
+1 (tag : <tag_name>)
+2 : 1 (<branch_name>)
+3
+4 : 1 (Example_child_of_1)
```
done

parkmichael777 commented on level_file_temp/setup.spec:
```diff
@@ -0,0 +1,9 @@
+2 : 1 (<branch_name>)
+3
+4 : 1 (Example_child_of_1)
+5 (tag : Example_child_of_4)
```
no capes

parkmichael777 commented on level_file_temp/test.spec:
```diff
@@ -0,0 +1,11 @@
+4 : 1 (Example_child_of_1)
+5 (tag : Example_child_of_4)
+6
+7 : 3 6 (Example_merge_node_of_3_&_4)
```
fixed

parkmichael777 commented on level_file_temp/test.spec:
```diff
@@ -0,0 +1,11 @@
+8
+-8 (Example_reverted_8)
+9 (Example_Changes_to_be_made_by_Player)
+10 (<checked_out_branch>)
```
fixed

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,140 @@
+    if not os.path.exists(os.path.join(skill_path,"__init__.py")):
+        with open(os.path.join(skill_path,"__init__.py"), 'w+') as fp:
+            level_setup = "\n".join([
+                "import pkg_resources\n",
```
done

parkmichael777 commented on setup.py:
```diff
@@ -33,9 +33,21 @@ def run(self):
+        'gitgud.skills.extras',
+        'gitgud.skills.rampup',
+        'git-gud.make_level.py',
+        'git-gud.level_file_temp'
```
done

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,140 @@
+        os.mkdir(level_path)
+        print("Created: {}".format(level_path))
+            
+    print()
```
done

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,140 @@
+            
+    print()
+    print("Creating Folders:")
+    print()
```
ok

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,140 @@
+    print()
+    write_init(skill_name,skill_path,level_name)
+            
+    print()
```
done ummm i think?

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,140 @@
+    create_level_file(level_path, "setup.spec")
+        
+    # Make test file
+    create_level_file(level_path, "test.spec")
```
done

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,140 @@
+                            "",
+                            "",
+                            "from gitgud.skills.u"
+                            ])
```
done

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,140 @@
+            
+    print()
+    print("Creating Files:")
+    print()
```
done

parkmichael777 commented:
Done with comments, pep8. Let me know if the logging looks good. If not, let me know what you want it to look like using an example format.

still need to do testcases

parkmichael777 commented:
test case file generation complete!

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,216 @@
+                "    simulate(gg, level, commands)"
+            ])
+            fp.write(test_setup)
+            print("Added entry for {}".format(level_name))
```
Added filepath

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,216 @@
+                "level_tests = [",
+                "    (",
+                "        skill[\'{}\'], [".format(level_name),
+                "            ",
```
Added example

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,216 @@
+                "",
+                "@pytest.mark.parametrize('level,commands', level_tests)",
+                "def test_level(gg, level, commands):",
+                "    simulate(gg, level, commands)"
```
Added

parkmichael777 commented on make_level.py:
```diff
@@ -0,0 +1,216 @@
+
+        # Add import statement
+        replace = "\n".join([
+            "from gitgud.skills.{0} import skil{0}_skill".format(skill_name),
```
Hmm not sure how that happened I must have accidentally pressed backspace without knowing where my cursor was lol

parkmichael777 commented on level_file_templates/test.spec:
```diff
@@ -0,0 +1,12 @@
+-8 (reverted-8)
+9 (changes-to-be-made-by-player)
+10 (checked-out-branch)
+checked-out-branch
```
noted

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+                "level_tests = [",
+                "    (",
+                "        skill[\'{}\'], [".format(level_name),
+                "            #Example: \'git gud commit\',",
```
Put #Example after this line

parkmichael777 replied:
done

benthayer replied:
I meant at the end. I just pushed the change I meant

benthayer commented:
In setup.py, the entry in package_data is missing a colon

benthayer commented:
Beyond that, most everything looks good. I'm moving onto reviewing the code itself

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+            "",
+            "from gitgud.skills.u"
+        ])
+        filedata = filedata.replace("\nfrom gitgud.skills.u", replace)
```
Change both of these lines to use the full line of `from gitgud.skills.util import AllSkills` leaving it cut off at the "u" makes it somewhat unclear what's happeneing under the hood.

Can you also change the name of the replace variable to `new_import`

parkmichael777 replied:
Done

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+        ])
+        filedata = filedata.replace("\nfrom gitgud.skills.u", replace)
+
+        # Add to input array of Skill
```
This comment isn't immediately clear. How about "Add skill to AllSkills"

parkmichael777 replied:
done

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+            fp.write(level_setup)
+
+        # Read file
+        with open(os.path.join("gitgud", "skills", "__init__.py"), 'r') as fp:
```
Add a comment stating that we're both adding an import statement and adding the new skill to AllSkills

parkmichael777 replied:
done

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+        with open(filepath, 'w') as fp:
+            fp.write(filedata)
+            print("Registered level \"{}\" in {}".format(level_name, filepath))
+    return
```
You don't need this return statement

parkmichael777 replied:
fixed

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+            print("Registered skill \"{}\" in {}".format(skill_name, filepath))
+
+        write_init(skill_name, skill_path, level_name)
+        return
```
Don't need this return statement

parkmichael777 replied:
fixed

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+        write_init(skill_name, skill_path, level_name)
+        return
+    else:
+        # Populate file with BasicLevel
```
Specify that it's the new skill file

parkmichael777 replied:
Done

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+            fp.write(filedata)
+            print("Registered skill \"{}\" in {}".format(skill_name, filepath))
+
+        write_init(skill_name, skill_path, level_name)
```
Since there's multiple `__init__.py`s you could be writing do, can you find a way to signal that this one is in the new skill's directory. That can either be by renaming the function or just a comment

parkmichael777 replied:
Hmmm, the write_init function writes both skill/__init__.py and skill/<new_skill>/__init__.py. The script adds to the former if the file doesn't exist, then the latter is written to recursively. If the file exists, then only the latter written to.

parkmichael777 replied:
I can keep the function name and then write a comment above it explaining this, or I can change the name; let me know.

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+        replace = ",\n        BasicLevel('{level_name}', pkg_resources.resource_filename(__name__, '_{level_name}/'))\n    ]".format(level_name=level_name)
+        filedata = filedata.replace("\n    ]", replace)
+        filedata = filedata.replace("[,", "[")
+        filepath = os.path.join(skill_path, "__init__.py")
```
Define and use this variable above since you use it when you open the file

parkmichael777 replied:
done

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+        replace = "\n".join([
+            "    ), (",
+            "        skill[\'{}\'], [".format(level_name),
+            "            #Example: \'git gud commit\',",
```
Can you change this to match what I did above?

parkmichael777 replied:
done

benthayer commented on make_level.py:
```diff
@@ -0,0 +1,217 @@
+                "@pytest.mark.parametrize('level,commands', level_tests)",
+                "def test_level(gg, level, commands):",
+                "    simulate(gg, level, commands)",
+                ""
```
What do you think about making a templates folder and moving this and the other init file into it? Then all we have to do is copy it and it'll look a lot cleaner. It'll also make make_level.py shorter and more managable

parkmichael777 replied:
done

benthayer commented:
> 
> 
> In setup.py, the entry in package_data is missing a colon

Thanks for getting this done, the file officially works perfectly!

benthayer commented on make_level.py:
```diff
@@ -47,49 +40,31 @@ def write_init(skill_name, skill_path, level_name):
-    return
 
 
 def write_test(skill_name, skill_path, level_name, level_path):
```
`skill_name` is not used, you can remove it as an argument

parkmichael777 replied:
done

benthayer commented on make_level.py:
```diff
@@ -141,38 +116,49 @@ def main():
-        "    package_data"
-    ]), replace1)
+    # Register new skill to setup.py
+    if not skill_exists:
```
It seems like we don't need this if statement or the `skill_exists` variable, we can just use the if statement from before

parkmichael777 replied:
removed

benthayer commented on make_level.py:
```diff
@@ -141,38 +116,49 @@ def main():
+        skill_exists = True
+
+    # Concatenate potential entry
     replace1 = "\n".join([
```
Indent both replaces, we only use them if the if statement executes

parkmichael777 replied:
Changed

benthayer commented on make_level.py:
```diff
@@ -141,38 +116,49 @@ def main():
+    while choice != 'y':
+        choice = input().lower()
+        if choice == 'n':
+            return
```
It might be useful to add `print("Aborting, no changes made.")`

parkmichael777 replied:
done

parkmichael777 commented:
New exception code and renamed folder.

parkmichael777 commented:
For test_levels.py, I made it a text file because if I leave it as a python file, CircleCI tries to compile it and it breaks because of the placeholder key. When it is actually used, it is renamed with a python format.

parkmichael777 commented on make_level.py:
```diff
@@ -92,19 +92,24 @@ def write_test(skill_name, skill_path, level_name, level_path):
     # Obtain input arguments
     try:
+        if (len(sys.argv) != 3):
+            raise IndexError(len(sys.argv))
```
If there are too little arguments an exception is raised, but not if there are too many.

benthayer replied:
Makes sense, but that's still not an index error. You can have multiple except clauses, so why not raise a generic `Exception` and handle it however you want

parkmichael777 replied:
Like do what I did before except with better logging? I couldn't find any better default exceptions besides IndexError and ValueError so I figured IndexError would be more applicable. Sorry, I've never had an opportunity to use try/except before this lol.

parkmichael777 replied:
pushed a new one, is that better?

benthayer replied:
I took a look and started working on figuring out the best way to do error handling and I eventually just wrote the code for simple if statements and pushed 😝

parkmichael777 commented:
Added ignore flag to circleci config

benthayer commented:
Nice, the ignore worked first try!

benthayer commented:
Touched up everything that I felt needed it and merged!
